miri/src/lib.rs

```diff
@@ -1,3 +1,789 @@
+                if old_size == 0 {
+                    return Err(EvalError::HeapAllocZeroBytes);
+                }
+                if !align.is_power_of_two() {
+                    return Err(EvalError::HeapAllocNonPowerOfTwoAlignment(align));
+                }
+                self.memory.deallocate(ptr, Some((old_size, align)), Kind::Rust)?;
+            }
+            "alloc::heap::::__rust_realloc" => {
+                let ptr = args[0].into_ptr(&mut self.memory)?.to_ptr()?;
+                let old_size = self.value_to_primval(args[1], usize)?.to_u64()?;
+                let old_align = self.value_to_primval(args[2], usize)?.to_u64()?;
+                let new_size = self.value_to_primval(args[3], usize)?.to_u64()?;
+                let new_align = self.value_to_primval(args[4], usize)?.to_u64()?;
+                if old_size == 0 || new_size == 0 {
+                    return Err(EvalError::HeapAllocZeroBytes);
+                }
+                if !old_align.is_power_of_two() {
+                    return Err(EvalError::HeapAllocNonPowerOfTwoAlignment(old_align));
+                }
+                if !new_align.is_power_of_two() {
+                    return Err(EvalError::HeapAllocNonPowerOfTwoAlignment(new_align));
+                }
+                let new_ptr = self.memory.reallocate(ptr, old_size, old_align, new_size, new_align, Kind::Rust)?;
+                self.write_primval(dest, PrimVal::Ptr(new_ptr), dest_ty)?;
+            }
+
+            // A Rust function is missing, which means we are running with MIR missing for libstd (or other dependencies).
+            // Still, we can make many things mostly work by "emulating" or ignoring some functions.
+            "std::io::_print" => {
+                trace!("Ignoring output.  To run programs that print, make sure you have a libstd with full MIR.");
+            }
+            "std::thread::Builder::new" => return Err(EvalError::Unimplemented("miri does not support threading".to_owned())),
+            "std::env::args" => return Err(EvalError::Unimplemented("miri does not support program arguments".to_owned())),
+            "std::panicking::panicking" |
+            "std::rt::panicking" => {
+                // we abort on panic -> `std::rt::panicking` always returns false
+                let bool = self.tcx.types.bool;
+                self.write_primval(dest, PrimVal::from_bool(false), bool)?;
+            }
+            _ => return Err(EvalError::NoMirFor(path)),
+        }
+
+        // Since we pushed no stack frame, the main loop will act
+        // as if the call just completed and it's returning to the
+        // current frame.
+        self.dump_local(dest);
+        self.goto_block(dest_block);
+        return Ok(());
+    }
+}
+
+//struct Memory<'b, 'a: 'b, 'tcx: 'a>(&'b rustc_miri::interpret::Memory<'a, 'tcx, Evaluator>);
+struct MemoryMut<'b, 'a: 'b, 'tcx: 'a>(&'b mut rustc_miri::interpret::Memory<'a, 'tcx, Evaluator>);
```

Comment:
Ok... I gave up on the newtype idea. It gets messy fast. We're back to extension traits. While those create some duplication of the function headers, they are so much more convenient to use.